### ISSUE TITLE
Molecule config files are interpreted as playbooks

### ISSUE THREAD

**Issue by antivirtel** [ansible-community/ansible-lint#1622]
##### Summary
We're using the latest Molecule (v3.3.4) to test our Ansible roles. Some roles require a full VM, so we're using Molecule tasks to create them. Ansible-lint latest thinks they're play files, but they should be considered as test files. Molecule puts their files to `<repo_root>/roles/<rolename>/molecule/<scenario:default>`. I know it should be under according to the collection layout: `<repo_root>/tests/molecule` or `<repo_root>/roles/<rolename>/tests/molecule/<scenario:default>`, but this is how it is generated still by molecule...

I took a look at `kind:` settings, but there's no role option. To skip this, I've used: `exclude_list:`. Unfortunately, it can't take wildcards, eg. `roles/*/molecule/**` so had to list all the roles with this.

The molecule itself will generate the files like `create_linux_instance.yml`. To generate an example file, install molecule, go to your roles dir (`<repo_root>/roles/<rolename>`), then execute:
`molecule init scenario default -d vmware`. 

Snippet (register is used multiple times after):
```yaml
- name: Create keypair
  user:
    name: "{{ lookup('env','USER') | default('root', true) }}"
    generate_ssh_key: true
    ssh_key_file: "{{ keypair_path }}"
  register: keypair_result
```

Error displayed:
```console
syntax-check: 'register' is not a valid attribute for a Play
roles/<rolename>/molecule/default/create_linux_instance.yml:2:3 ERROR! 'register' is not a valid attribute for a Play

The error appears to be in '<repo_root>/roles/<rolename>/molecule/default/create_linux_instance.yml': line 2, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

---
- name: Create keypair
  ^ here
```

##### Issue Type

- Bug Report

##### Ansible and Ansible Lint details
```console (paste below)
$ ansible --version
ansible 2.9.22
  config file = ~/deploy/<repo_root>/ansible.cfg
  configured module search path = ['~/deploy/<repo_root>/library']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Nov 16 2020, 16:55:22) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]

$ ansible-lint --version
ansible-lint 5.0.12 using ansible 2.9.22

```

- ansible installation method: one of source, pip, OS package: pip
- ansible-lint installation method: one of source, pip, OS package: pip

##### OS / ENVIRONMENT
$ uname -a
Linux <hostname> 3.10.0-1127.18.2.el7.x86_64 #1 SMP Sun Jul 26 15:27:06 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

$ cat /etc/redhat-release
CentOS Linux release 7.8.2003 (Core)

##### STEPS TO REPRODUCE
To generate an example file, install molecule, go to any roles dir (`<repo_root>/roles/<rolename>`), then execute:
`molecule init scenario default -d vmware`. 

Snippet (register is used multiple times after):
```yaml
- name: Create keypair
  user:
    name: "{{ lookup('env','USER') | default('root', true) }}"
    generate_ssh_key: true
    ssh_key_file: "{{ keypair_path }}"
  register: keypair_result
```

##### Desired Behaviour
Molecule files interpreted as test files, not playbooks.

##### Actual Behaviour
```yaml
# the only way of bypassing this error.
#exclude_paths:
# - 'roles/<rolename>/molecule/default'

skip_list:
 - 'yaml'
 - '204' # Lines <= 160
 - 'var-spacing'
 - '208' # file permissions unset
 - 'risky-file-permissions'
 - '403' # package install should not use latest
 - 'package-latest'
 - '501' # become user requires become
 - 'partial-become'
 - '602' # don't compare to empty string
 - 'empty-string-compare'
 - '701' # meta/main should contain: author, description, license, min_ansible_version, platforms
 - 'meta-no-info'
 - 'experimental'

# syntax-check will fail without defined vars like this
extra_vars:
  user_defined_hosts: ['localhost']
  adhoc_ip_address: '127.0.0.1'
  inventory_target: 'localhost'

verbosity: 0
```

```console
$ ansible-lint -vvvv -c .ansible-lint roles
DEBUG    Logging initialized to level 10
DEBUG    Options: Namespace(cache_dir='~/.cache/ansible-lint/ccc58c', colored=True, config_file='~/deploy/<rep_root>/.ansible-lint', configured=True, cwd=PosixPath('~/deploy/<rep_root>'), display_relative_path=True, enable_list=[], exclude_paths=['.cache', '.git', '.hg', '.svn', '.tox'], extra_vars={'user_defined_hosts': ['localhost'], 'adhoc_ip_address': '127.0.0.1', 'inventory_target': 'localhost'}, format='rich', kinds=[{'jinja2': '**/*.j2'}, {'jinja2': '**/*.j2.*'}, {'requirements': '**/meta/requirements.yml'}, {'galaxy': '**/galaxy.yml'}, {'reno': '**/releasenotes/*/*.{yaml,yml}'}, {'playbook': '**/playbooks/*.{yml,yaml}'}, {'playbook': '**/*playbook*.{yml,yaml}'}, {'role': '**/roles/*/'}, {'tasks': '**/tasks/**/*.{yaml,yml}'}, {'handlers': '**/handlers/*.{yaml,yml}'}, {'vars': '**/{host_vars,group_vars,vars,defaults}/**/*.{yaml,yml}'}, {'meta': '**/meta/main.{yaml,yml}'}, {'yaml': '.config/molecule/config.{yaml,yml}'}, {'requirements': '**/molecule/*/{collections,requirements}.{yaml,yml}'}, {'yaml': '**/molecule/*/{base,molecule}.{yaml,yml}'}, {'requirements': '**/requirements.yml'}, {'playbook': '**/molecule/*/*.{yaml,yml}'}, {'yaml': '**/{.ansible-lint,.yamllint}'}, {'yaml': '**/*.{yaml,yml}'}, {'yaml': '**/.*.{yaml,yml}'}], lintables=['roles'], listrules=False, listtags=False, loop_var_prefix=None, mock_modules=[], mock_roles=[], offline=False, parseable=False, parseable_severity=False, progressive=False, project_dir='.', quiet=False, rules={}, rulesdir=[], rulesdirs=['/usr/local/lib/python3.6/site-packages/ansiblelint/rules'], skip_action_validation=True, skip_list=['yaml', '204', 'var-spacing', 'risky-file-permissions', 'risky-file-permissions', 'package-latest', 'package-latest', 'partial-become', 'partial-become', 'empty-string-compare', 'empty-string-compare', 'meta-no-info', 'meta-no-info', 'experimental'], tags=[], use_default_rules=False, var_naming_pattern='^[a-z_][a-z0-9_]*$', verbosity=4, version=False, warn_list=['experimental', 'role-name'])
INFO     Added ANSIBLE_ROLES_PATH=~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles:roles:roles
DEBUG    Loading rules from /usr/local/lib/python3.6/site-packages/ansiblelint/rules
INFO     Added ANSIBLE_ROLES_PATH=~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles:roles:roles
Loading custom .yamllint config file, this extends our internal yamllint config.
INFO     Discovered files to lint using: git ls-files -z
INFO     Executing syntax check on roles/machine_base/molecule/default/create_linux_instance.yml (3.03s)
[... a long list of files here...]
INFO     Executing syntax check on roles/grafana/molecule/default/converge.yml (3.44s)
WARNING  Listing 4 violation(s) that are fatal
syntax-check: 'register' is not a valid attribute for a Play
roles/<rolename>/molecule/default/create_linux_instance.yml:2:3 ERROR! 'register' is not a valid attribute for a Play

The error appears to be in '~/deploy/<rep_root>/roles/docker/molecule/default/create_linux_instance.yml': line 2, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

---
- name: Create keypair
  ^ here

[... 3 more ...]
```

**Comment by tadeboro** [comment]
This looks like a "problem" with the `molecule-vmware` driver that does not place task files into a `tasks` directory and instead places them in the root of the scenario.

If my assumptions are correct, you can bypass this problem by moving the `create_linux_instance.yml`  and `create_windows_instance.yml` files into a `tasks` directory inside your scenarios. This way, the linter will properly classify them as ask files and lint them accordingly.

Once we know this works as expected, we should open a bug report or create a PR in the https://github.com/sky-joker/molecule-vmware/ repo and update the cookie-cutter template that is used when initializing new scenarios (https://github.com/sky-joker/molecule-vmware/tree/master/molecule_vmware/cookiecutter/%7B%7Bcookiecutter.molecule_directory%7D%7D/%7B%7Bcookiecutter.scenario_name%7D%7D).

**Comment by antivirtel** [comment]
Ok, I've opened an issue: https://github.com/sky-joker/molecule-vmware/issues/20

As of https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_structure.html#collection-directories-and-files -> I think the whole thing might need to be under <repo_root>/roles/<rolename>/tests/molecule, doesn't it?

**Comment by tadeboro** [comment]
Ansible itself knows nothing about Molecule because this is an external tool not developed by the core team. So searching official docs will give you back zero results.

Let me try to explain with an example how things look right now and what changes I am suggesting. So, after we run the `molecule init scenario default -d vmware` command, we end up with this directory structure:

```
$ tree
roles/
└── my.role
    ├── defaults
    │   └── main.yml
    ├── files
    ├── handlers
    │   └── main.yml
    ├── meta
    │   └── main.yml
    ├── molecule
    │   └── default
    │       ├── converge.yml
    │       ├── create_linux_instance.yml
    │       ├── create_windows_instance.yml
    │       ├── create.yml
    │       ├── destroy.yml
    │       ├── INSTALL.rst
    │       ├── molecule.yml
    │       └── verify.yml
    ├── README.md
    ├── tasks
    │   └── main.yml
    ├── templates
    ├── tests
    │   ├── inventory
    │   └── test.yml
    └── vars
        └── main.yml
```

`ansible-lint` "knows" that things that Molecule stores inside the `roles/my.role/molecule/default/` directory are playbooks, not task files. But the VMware driver breaks that assumption because the `create_(linux|windows)_instance.yml` files are not playbooks, they are task files.

The change I proposed is the following one:

```
$ mkdir roles/my.role/molecule/default/tasks
$ mv roles/my.role/molecule/default/create_* roles/my.role/molecule/default/tasks
$ tree
roles/
└── my.role
    ├── defaults
    │   └── main.yml
    ├── files
    ├── handlers
    │   └── main.yml
    ├── meta
    │   └── main.yml
    ├── molecule
    │   └── default
    │       ├── converge.yml
    │       ├── create.yml
    │       ├── destroy.yml
    │       ├── INSTALL.rst
    │       ├── molecule.yml
    │       ├── tasks
    │       │   ├── create_linux_instance.yml
    │       │   └── create_windows_instance.yml
    │       └── verify.yml
    ├── README.md
    ├── tasks
    │   └── main.yml
    ├── templates
    ├── tests
    │   ├── inventory
    │   └── test.yml
    └── vars
        └── main.yml
```

So, no files were actually moved outside the scenario directory, we just partitioned them according to the standard Ansible practices.

@antivirtel Can you please confirm that this change works for you? Once we have that confirmation, we can work on updating the molecule-vmware package.

**Comment by antivirtel** [comment]
Thanks, we're trying to do that, but it doesn't seem like `ansible-lint` accepts it, see: https://github.com/sky-joker/molecule-vmware/issues/20#issuecomment-861653355 - any suggestions?

**Comment by ssbarnea** [comment]
Closing as that is not really a bug with the linter. In fact the linter did a great job here on identifying bad code layout. The only non-playbook inside `molecule/` folders is `molecule.yml` file, everything else is assumed to be a valid playbook. If you have task files, you must put them inside `tasks/` subfolders. If you have other kind of random yml files, you must put them in subfolders, like `files/` or whatever you find suitable. @tadeboro made a really good graphical representation of a good layout.

@sky-joker/  At least today I learnt that there is a molecule-vmware driver. Nice work. Just keep in mind that you can move the repo under @ansible-community organization and still keep the ownership on your driver (it can make it easier to discover by others).

**Comment by sky-joker** [comment]
Hi @ssbarnea 

> @sky-joker At least today I learnt that there is a molecule-vmware driver. Nice work. Just keep in mind that you can move the repo under @ansible-community organization and still keep the ownership on your driver (it can make it easier to discover by others).

Thank you for telling me that.
How do I transfer molecul-vmware to under the community organization?

**Comment by ssbarnea** [comment]
Transfer it to me and I will move it into community right away as the user making the transfer needs extra permissions in target org.

**Comment by sky-joker** [comment]
ok!
I'm working on fixing this issue, so I'll transfer molecule-vmware to you after it is resolved.

**Comment by sky-joker** [comment]
@ssbarnea
I have transferred molecule-vmware to you.
Could you please check it?

**Comment by ssbarnea** [comment]
Yep did it and also invited you to be repo admin. Check your email.

**Comment by sky-joker** [comment]
Thanks, @ssbarnea!
I confirmed that molecule-vmware has transferred to the community.

**Comment by ssbarnea** [comment]
@geoffreyvanwyk I think that you are confused by a feature of ansible-lint that was there from day one: any file argument passed to it is assumed to be a playbook, regardless of what patterns you configured for it. Did you read the warning line about "overriding detected file kind"? That was supposed to tell you what was happening.

We had discussions about removing this feature, but we may not be able to make it without bumping the major version, as it counts as not not backwards compatible.

@cidrblock @ganeshrn Maybe it time to bite the bullet and change that, I postponed it for years and I see that it still confuses users.
